### PR TITLE
Print an error message if memory allocation fails

### PR DESCRIPTION
**Pull request by wthrowe**
Switches back to `new` from `malloc` because `new` has support for this built in and does not seem to be significantly slower when the version that does not zero the memory is used.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by nilsdeppe — COMMENTED** [review]
I'd be fine with this in debug mode, not release mode.

**Comment by wthrowe** [comment]
Fine with what in particular?

**Comment by nilsdeppe** [comment]
I'm a bit worried about the performance hit in release mode, at least currently.

I thought about this more and do you think having a custom allocator for use with STL containers would be a good/better approach? Specifically, I do find all the `unique_ptr` things a bit tedious and when I looked a month or so ago we were basically just using them as `std::vector`s with a custom allocator (`malloc`). We could then do a couple things in the allocator, one being use the approach you did here, another being enabling an OOM check with malloc (possibly only in debug mode, but that can be discussed once we've figured out whether to do the allocator thing at all). What are your thoughts on using a fast allocator and then `std::vector`?

I want to note that I'm not suggesting this be changed in this PR or by you, I just would like to get your feedback on the idea as an alternative that might also help give error messages and be useful in other containers :)

**Comment by wthrowe** [comment]
The overhead compared to `malloc` was something like 2% on a microbenchmark of the allocator.  I don't think that's a significant performance hit.

The default allocator is just a wrapper around `malloc` with an OOM check, so I'm not sure how we could write a better one without doing something really complicated.

The change to print errors affects all the standard containers (because they use `new`).  The other commit was necessary to have it also affect `DataVector` etc.

**Comment by nilsdeppe** [comment]
Okay, I think I've understood my source of confusion. I'll do some reading and tests to make sure I understand. If my understanding of this is now correct I'm fine with the changes (I'll need to do a detailed review for typos, etc. still) :)

File: src/DataStructures/VectorImpl.hpp [hunk 1]
```diff
@@ -368,9 +370,10 @@ void VectorImpl<T, VectorType>::pup(PUP::er& p) noexcept {  // NOLINT
   if (my_size > 0) {
     if (p.isUnpacking()) {
       owning_ = true;
-      owned_data_.reset(my_size > 0 ? static_cast<value_type*>(
-                                          malloc(my_size * sizeof(value_type)))
-                                    : nullptr);
+      owned_data_.reset();
+      if (my_size > 0) {
```
Comment (moxcodes) [inline src/DataStructures/VectorImpl.hpp]: isn't this if redundant with the one on line 370?

**Comment by wthrowe** [comment]
Rebased.

File: src/Utilities/MemoryHelpers.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+struct unique_type<T[]> {
+  using array = std::unique_ptr<T[]>;
+};
+
+template<typename T, size_t Bound>
+struct unique_type<T[Bound]> { struct invalid_type { }; };
+}  // namespace detail
+
+template <typename T, typename... Args>
+typename detail::unique_type<T>::single_object make_unique_default_init() {
```
Comment (nilsdeppe) [inline src/Utilities/MemoryHelpers.hpp]: I tried to find this on cpp reference and it looks like this was rename to `make_unique_for_overwrite` (https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique). I'm happy to leave this as an issue to do a find-replace on in the future in the interest of getting this merged sooner :)

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@wthrowe FYI: CI is unhappy with: `MemoryHelpers.hpp should be added to ./src/Utilities/CMakeLists.txt`

File: src/DataStructures/Variables.hpp [hunk 3]
```diff
@@ -72,12 +72,12 @@ class Variables;
...
  * memory allocations are quite expensive, especially in a parallel environment.
  *
- * `Variables` stores the data it owns in a `std::unique_ptr<double[],
- * decltype(&free)>` instead of a `std::vector` because allocating the
- * `unique_ptr` with `malloc` allows us to avoid initializing the memory
- * completely in release mode when no value is passed to the constructor.
- * Additionally, if the macro `SPECTRE_NAN_INIT` is defined, initialization with
- * `NaN`s is done even in release mode.
+ * In Debug mode, or if the macro `SPECTRE_NAN_INIT` is defined, the contents
+ * of are initialized with `NaN`s.
```
Comment (kidder) [inline src/DataStructures/Variables.hpp]: of are -> are

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

**Review by moxcodes — DISMISSED** [review]

File: src/Utilities/MemoryHelpers.hpp (hunk 2 above)
Comment (wthrowe) [reply src/Utilities/MemoryHelpers.hpp]: Running sed is easy and I had other changes to make anyway.

**Review by wthrowe — COMMENTED** [review]
I think all the clang-tidy stuff is charm++ internals, but if I missed something let me know.

**Review by nilsdeppe — DISMISSED** [review]

**Review by moxcodes — DISMISSED** [review]

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
ignoring clang-tidy error about charm++

**Comment by kidder** [comment]
sorry @wthrowe forgot to push the merge button after it finished CI and now this needs a rebase

**Review by moxcodes — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**